### PR TITLE
[quick_actions]Migrate the plugin class to Swift, and remove custom modulemap

### PR DESCRIPTION
**Pull request by hellohuanlin**
As [suggested](https://github.com/flutter/plugins/pull/6229#discussion_r943872835) in https://github.com/flutter/plugins/pull/6229, this PR contains 2 changes together: (1) migration of `FLTQuickActionsPlugin` class, and (2) removing custom modulemap 

# About migration of `FLTQuickActionsPlugin`: 

## Why migrating this class first

This is following the "top down approach" discussed in [the proposal](https://docs.google.com/document/d/1XsaulkJA6_ZSpM7chkQLhQY25sQqhEMiqHMYzw8H85o/edit?resourcekey=0-_cUjF1c0iBvRLKfV-3gK5A#heading=h.reauq7bq8vak). The migration is pretty smooth - Swift has much better support when importing ObjC. 

## Why test is not migrated in this PR

Migrating tests in separate PRs will give us better confidence, because the code will be verified by both ObjC and Swift tests, according to [the proposal](https://docs.google.com/document/d/1XsaulkJA6_ZSpM7chkQLhQY25sQqhEMiqHMYzw8H85o/edit?resourcekey=0-_cUjF1c0iBvRLKfV-3gK5A#heading=h.q26k0rzfcant): 

> "For ObjC code that is already tested, make sure to also write tests in Swift. **It is even better if the existing ObjC test just works for the migrated Swift code**, because it provides us with good confidence that the new Swift code behaves exactly the same as the original ObjC code. Though due to interoperability and OCMock limitations discussed above, it is not always possible to migrate code and its test in separate steps. "

## Swift format

For now I run [swift-format](https://github.com/apple/swift-format) on my local machine. But we should add the check to CI in the future, before wider community starts contributing Swift code. 

# About removing custom modulemap: 

I do not like this solution since it makes private headers public. However, I wasn't able to find a better solution after lots of research on this topic. 

Below is a summary of what I found so far, in case it's useful searching for a better solution in the future: 

1. Cocoapods does not support a custom modulemap file for static libraries that contains Swift: 

>     [!] Using Swift static libraries with custom module maps is currently not supported. Please build `quick_actions_ios` as a framework or remove the custom module map.

**This limitation is the culprit of everything**. If we could use a custom modulemap, we could easily define a submodule (aka we are currently doing) or use a private modulemap to map the private headers, [like this](https://stackoverflow.com/questions/28746214/how-to-import-private-framework-headers-in-a-swift-framework). 

2. Adding `use_frameworks!` to the `Podfile` also works, but we can't force developers to choose frameworks and not static libraries. 

3. There seems to be some effort fixing this problem in Cocoapods ([this](https://github.com/CocoaPods/CocoaPods/pull/9363) and [this](https://github.com/CocoaPods/CocoaPods/issues/4656)), but there isn't much update since then. 

4. What about removing clang modules at all? In pure ObjC it's fine, but Swift cannot directly import headers - it can only interact with modules. 

5. The auto-generated default modulemap simply exports everything from the umbrella header. It defines a submodule for each of the header import, and re-export them to the main module. Only public headers are allowed in this process. Something like this: 

```
module ModuleName {
  umbrella header "ModuleName-umbrella.h"
  export * 
  module * { export * }
}
```

This is only a temporary workaround during migration, but the bad state could last a long time for larger plugins. Though on the bright side, consider that [private headers aren't that "private" at all](https://stackoverflow.com/questions/7439192/xcode-copy-headers-public-vs-private-vs-project) - they can be imported by clients just like public headers. it's only that they are put under a separate directory named "PrivateHeaders" to warn the clients that it's risky to import them, but nothing can technically stop them from doing so. 

*List which issues are fixed by this PR. You must list at least one issue.*

https://github.com/flutter/flutter/issues/108750

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    if let shortcutItem = launchOptions[UIApplication.LaunchOptionsKey.shortcutItem] as? UIApplicationShortcutItem {
+      // Keep hold of the shortcut type and handle it in the
+      // `applicationDidBecomeActure:` method once the Dart MethodChannel
+      // is initialized.
+      launchingShortcutType = shortcutItem.type
+
+      // Return false to indicate we handled the quick action to ensure
+      // the `application:performActionFor:` method is not called (as
+      // per Apple's documentation:
+      // https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622935-application?language=objc).
```
Comment (hellohuanlin) [inline packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift:71]: the comments are simply copied over from objc.

File: packages/quick_actions/quick_actions_ios/ios/Classes/quick_actions_ios-umbrella.h [hunk 2]
```diff
@@ -7,4 +7,4 @@
 FOUNDATION_EXPORT double quickActionsIOSVersionNumber;
 FOUNDATION_EXPORT const unsigned char quickActionsIOSVersionString[];
 
-#import <quick_actions_ios/FLTQuickActionsPlugin.h>
+#import <quick_actions_ios/FLTShortcutStateManager.h>
```
Comment (hellohuanlin) [inline packages/quick_actions/quick_actions_ios/ios/Classes/quick_actions_ios-umbrella.h:10]: Since the custom module map is gone, this has to be here, until the migration is done (refer to the description of this PR)

File: packages/quick_actions/quick_actions_ios/ios/quick_actions_ios.podspec [hunk 3]
```diff
@@ -14,11 +14,14 @@ Downloaded by pub (not CocoaPods).
...
-  s.source_files = 'Classes/**/*.{h,m}'
-  s.public_header_files = 'Classes/*.h'
-  s.private_header_files = 'Classes/PrivateHeaders/*.h'
+  s.swift_version = '5.0'
+  s.source_files = 'Classes/**/*.{h,m,swift}'
+  s.xcconfig = {
+     'LIBRARY_SEARCH_PATHS' => '$(TOOLCHAIN_DIR)/usr/lib/swift/$(PLATFORM_NAME)/ $(SDKROOT)/usr/lib/swift',
+     'LD_RUNPATH_SEARCH_PATHS' => '/usr/lib/swift',
+  }
+  s.public_header_files = 'Classes/**/*.h'
```
Comment (hellohuanlin) [inline packages/quick_actions/quick_actions_ios/ios/quick_actions_ios.podspec:23]: unfortunately we have to expose private headers until migration is complete (refer to the PR description)

File: packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  ) {
+    self.channel = channel
+    self.shortcutStateManager = shortcutStateManager
+  }
+
+  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
+    switch call.method {
+    case "setShortcutItems":
+      // `arguments` must be an array of dictionaries
+      let items = call.arguments as! [[String: Any]]
```
Comment (hellohuanlin) [inline packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift:37]: Force unwrap it to keep the same behavior as ObjC. Whether to crash on platform code or to surface an error is discussed [here in this proposal](https://docs.google.com/document/d/1Iqt36mp_MWX7Ou2l9tQpUPjlNbh0rSRrni7X0DZK3pk/edit?resourcekey=0-XdZ6l6uykkhbBbivPl64lg#heading=h.1yb3rbn1zp2z).

File: packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    {
+      // Keep hold of the shortcut type and handle it in the
+      // `applicationDidBecomeActure:` method once the Dart MethodChannel
+      // is initialized.
+      launchingShortcutType = shortcutItem.type
+
+      // Return false to indicate we handled the quick action to ensure
+      // the `application:performActionFor:` method is not called (as
+      // per Apple's documentation:
+      // https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622935-application).
```
Comment (hellohuanlin) [inline packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift:73]: these comments are copied over from objc.

File: packages/quick_actions/quick_actions_ios/ios/Classes/FLTQuickActionsPlugin.m [hunk 6]
```diff
@@ -1,99 +0,0 @@
...
-  } else if ([call.method isEqualToString:@"getLaunchAction"]) {
-    result(nil);
-  } else {
-    result(FlutterMethodNotImplemented);
-  }
-}
-
-- (void)dealloc {
-  [_channel setMethodCallHandler:nil];
-  _channel = nil;
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/quick_actions/quick_actions_ios/ios/Classes/FLTQuickActionsPlugin.m:56]: > Also double checked other plugins and they don't call `setMethodCallHandler:nil`

It's been a while since I dug into plugin lifetime, but IIRC it's the channel handler registration that keeps the plugin instance alive in the first place; if that's true, then this will never do anything in `dealloc` since we couldn't be here if there's still an active channel. I may be misremembering though.

File: packages/quick_actions/quick_actions_ios/pubspec.yaml [hunk 7]
```diff
@@ -13,7 +13,7 @@ flutter:
     implements: quick_actions
     platforms:
       ios:
-        pluginClass: FLTQuickActionsPlugin
+        pluginClass: QuickActionsPlugin
```
Comment (stuartmorgan) [inline packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: Don't we still require the thin ObjC wrapper around Swift plugins? That's still what we are generating in the template; I thought it was required for the Swift-plugin-in-ObjC host to work in iOS apps. @jmagman

File: packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+  public func application(
+    _ application: UIApplication,
+    didFinishLaunchingWithOptions launchOptions: [AnyHashable: Any] = [:]
+  ) -> Bool {
+    if let shortcutItem = launchOptions[UIApplication.LaunchOptionsKey.shortcutItem]
+      as? UIApplicationShortcutItem
+    {
+      // Keep hold of the shortcut type and handle it in the
+      // `applicationDidBecomeActure:` method once the Dart MethodChannel
```
Comment (stuartmorgan) [inline packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift:66]: This was presumably meant to say "Active", not "Acture"; we shouldn't preserve the typo.

File: packages/quick_actions/quick_actions_ios/ios/Classes/FLTQuickActionsPlugin.m (hunk 6 above)
[the remark this answers is not in the document]
Comment (hellohuanlin) [reply packages/quick_actions/quick_actions_ios/ios/Classes/FLTQuickActionsPlugin.m:56]: Gotcha. This would be a serious memory issue for our plugins. I created https://github.com/flutter/flutter/issues/111206 to track it.

File: packages/quick_actions/quick_actions_ios/pubspec.yaml (hunk 7 above)
Comment (hellohuanlin) [reply packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: The generated file is this: 

```
#import "GeneratedPluginRegistrant.h"

#if __has_include(<integration_test/IntegrationTestPlugin.h>)
#import <integration_test/IntegrationTestPlugin.h>
#else
@import integration_test;
#endif

#if __has_include(<quick_actions_ios/QuickActionsPlugin.h>)
#import <quick_actions_ios/QuickActionsPlugin.h>
#else
@import quick_actions_ios;
#endif

@implementation GeneratedPluginRegistrant

+ (void)registerWithRegistry:(NSObject<FlutterPluginRegistry>*)registry {
  [IntegrationTestPlugin registerWithRegistrar:[registry registrarForPlugin:@"IntegrationTestPlugin"]];
  [QuickActionsPlugin registerWithRegistrar:[registry registrarForPlugin:@"QuickActionsPlugin"]];
}

@end
```

It is directly using `[QuickActionsPlugin registerWithRegistrar:]`. So we should be fine here?

Comment (stuartmorgan) [reply packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: @jmagman Did `__has_include` resolve all the issues with direct Swift-into-Obj-C import? If so, we should fix the plugin template to stop generating the Obj-C wrapper.

Comment (hellohuanlin) [reply packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: Are you suggesting to remove `GeneratedPluginRegistrant.h/m` and move the logic inside the `AppDelegate`?

Comment (stuartmorgan) [reply packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: No, I'm talking about the ObjC files [here](https://github.com/flutter/flutter/tree/master/packages/flutter_tools/templates/plugin/ios-swift.tmpl/Classes).

It seems like either there is still at least one combination of app language and library build type that requires those wrappers, in which case this plugin needs them too, or that's no longer the case, in which case the template shouldn't need them.

Comment (hellohuanlin) [reply packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: Gotcha. I did not find any clue about why we needed this wrapper from [the original PR](https://github.com/flutter/flutter/pull/10259)

The wrapped swift class is an NSObject subclass which is already exposed to objc runtime.  

- objc plugin + objc/swift app: unchanged
- swift plugin + objc app: `GeneratedPluginRegistrant.m` directly uses `QuickActionsPlugin.swift` as shown in this example
- swift plugin + swift app: shouldn't involve any objc logic at all. 

So I don't think the ObjC wrapper is necessary.

Comment (stuartmorgan) [reply packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: @jmagman Did extensive investigation a while back about various combinations of host app language, plugin language, and compilation mode (static library vs dynamic library), so should know if there are still edge cases that require it or not.

File: packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift (hunk 4 above)
Comment (hellohuanlin) [reply packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift:37]: If wrong type it should crash: 

```
NSNumber *arr = @1;
for (NSDictionary *item in (NSArray *)arr) { // crash here
  NSLog(@"%@", item);
}
```
But I see your point about `nil` value. The equivalent swift should be this instead: 

```
if let arguments = call.arguments {
  // crash of wrong type
  items = arguments as! [[String:Any]]
} else {
  // no crash if nil
  items = []
}
```

Not sure if it's worth to do tho. The corresponding dart side guarantees that the argument is non-null. So the `else` block actually never run.

Comment (cyanglaz) [reply packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift:37]: SGTM

**Review by cyanglaz — APPROVED** [review]
Swift code LGTM

File: packages/quick_actions/quick_actions_ios/pubspec.yaml (hunk 7 above)
Comment (jmagman) [reply packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: Objc-plugin-in-Swift-app will work without the wrapper if the Objective-C plugins define modules correctly, which they should out of the box as of plugin podspec template change in https://github.com/flutter/flutter/pull/40302 (Flutter 1.10.3).  Plugins created before that point may not work when used in a Swift app since it's missing `DEFINES_MODULE`.  I think all other combinations should work without the wrapper.

We could add a warning to the tool for the plugin author that they need to add `{ 'DEFINES_MODULE' => 'YES' }` to their podspec (although I don't think we have a good tool hook to know they are building a plugin example app and not just a regular app).  Or we could add some documentation about how to fix up the podspec when their plugin customers report it doesn't work in a Swift app.

**Review by jmagman — APPROVED** [review]
LGTM

File: packages/quick_actions/quick_actions_ios/pubspec.yaml (hunk 7 above)
Comment (hellohuanlin) [reply packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: @jmagman thanks for the explanation. Sounds like we should remove it. Created an issue here https://github.com/flutter/flutter/issues/111928

Comment (jmagman) [reply packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: > Are all the languages swapped in this comment by accident, or we talking about different cases? I was talking about the ObjC wrapper that we generate in the Swift plugin template, to make them look like older (ObjC) plugins, which IIRC was necessary for the Swift-plugin-in-ObjC-app case (for at least one of the library build modes).

You're right I was mixing this up with the Objective-C files that get generated in a Swift iOS app, which I think was originally so the Swift app can import Objective-C plugins pre-modules.
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner/Runner-Bridging-Header.h
https://github.com/flutter/flutter/blob/5aad12cab1e9c0863ea5fbc9a770b2734d7772db/packages/flutter_tools/lib/src/flutter_plugins.dart#L456

Screen Shot 2022-09-19 at 3 58 35 PM

But you're right, I _think_ the Swift plugins also don't need the Objective-C code now.  Fortunately this area is pretty well tested, so if someone tries https://github.com/flutter/flutter/issues/111928 the tests should let us know if there are any corner cases I'm not remembering.

**Comment by jmagman** [comment]
@hellohuanlin is this ready to land?

**Comment by hellohuanlin** [comment]
@jmagman sorry i was busy with the video player ios 16 bug. will rebase it to fix the submit-queue and land later.

File: packages/quick_actions/quick_actions_ios/pubspec.yaml (hunk 7 above)
Comment (hellohuanlin) [reply packages/quick_actions/quick_actions_ios/pubspec.yaml:16]: Reading it again (@jmagman you gave me hints about clang module), i think it's possible that the original plugin template wanted to expose both header and module, so that its objc client can freely choose from importing headers vs importing modules. This benefit seems so trivial that we should probably remove the template.
